pkg/image/controller/trigger/image_trigger_controller.go

```diff
@@ -0,0 +1,437 @@
+		lister:           lister,
+		tagRetriever:     tagRetriever{lister},
+		triggerCache:     NewTriggerCache(),
+
+		resourceFailureDelayFn: defaultResourceFailureDelay,
+	}
+
+	c.syncImageStreamFn = c.syncImageStream
+	c.syncResourceFn = c.syncResource
+	c.enqueueImageStreamFn = c.enqueueImageStream
+
+	isInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addImageStreamNotification,
+		UpdateFunc: c.updateImageStreamNotification,
+	})
+	c.syncs = []cache.InformerSynced{isInformer.HasSynced}
+
+	triggers, syncs, err := setupTriggerSources(c.triggerCache, c.tagRetriever, sources, c.imageChangeQueue)
+	if err != nil {
+		panic(err)
+	}
+	c.triggerSources = triggers
+	c.syncs = append(c.syncs, syncs...)
+
+	return c
+}
+
+// setupTriggerSources is used by test code to simulate a trigger controller.
+func setupTriggerSources(triggerCache cache.ThreadSafeStore, tagRetriever trigger.TagRetriever, sources []TriggerSource, imageChangeQueue workqueue.RateLimitingInterface) (map[string]TriggerSource, []cache.InformerSynced, error) {
+	var syncs []cache.InformerSynced
+	triggerSources := make(map[string]TriggerSource)
+	for _, source := range sources {
+		if source.Store == nil {
+			source.Store = source.Informer.GetStore()
+		}
+		prefix := source.Resource.String() + "/"
+		if _, ok := triggerSources[source.Resource.String()]; ok {
+			return nil, nil, fmt.Errorf("duplicate resource names registered in %#v", sources)
+		}
+		triggerSources[source.Resource.String()] = source
+
+		handler := ProcessEvents(triggerCache, source.TriggerFn(prefix), imageChangeQueue, tagRetriever)
+		source.Informer.AddEventHandler(handler)
+		syncs = append(syncs, source.Informer.HasSynced)
+	}
+	return triggerSources, syncs, nil
+}
+
+// Run begins watching and syncing.
+func (c *TriggerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting trigger controller")
+
+	if !cache.WaitForCacheSync(stopCh, c.syncs...) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.imageStreamWorker, time.Second, stopCh)
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.resourceWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down trigger controller")
+}
+
+func (c *TriggerController) addImageStreamNotification(obj interface{}) {
+	is := obj.(*imageapi.ImageStream)
+	c.enqueueImageStreamFn(is)
+}
+
+func (c *TriggerController) updateImageStreamNotification(old, cur interface{}) {
+	c.enqueueImageStreamFn(cur.(*imageapi.ImageStream))
+}
+
+func (c *TriggerController) enqueueImageStream(is *imageapi.ImageStream) {
+	key, err := controller.KeyFunc(is)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", is, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *TriggerController) enqueueRateLimited(is *imageapi.ImageStream) {
```

Comment:
Actually this is the thing we've discussed yesterday during our arch call, wrt to rate limiting the resources that can be triggered at once. Eg. when a base image changes we don't want to trigger all builds and deployments at once and ddos the cluster. It can be addressed as a followup, but it's an important topic. See https://github.com/openshift/origin/issues/13487 for reference.